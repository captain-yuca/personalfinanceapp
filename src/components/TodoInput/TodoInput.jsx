import React, { useState } from 'react'
import uniqid from 'uniqid'
import PropTypes from 'prop-types'

import { Wrapper, Input, Button } from './TodoInput.styles'

const propTypes = {
  addTodo: PropTypes.func,
}

function TodoInput({ addTodo }) {
  const [value, setValue] = useState('')

  const handleAddTodo = () => {
    if (!value) return
    addTodo({ text: value, id: uniqid() })
    setValue('')
  }

  return (
    <Wrapper>
      <Input placeholder="Enter new todo" type="text" value={value} onChange={e => setValue(e.target.value)} />
      <Button onClick={handleAddTodo}>Add</Button>
    </Wrapper>
  )
}

TodoInput.propTypes = propTypes

export default TodoInput
