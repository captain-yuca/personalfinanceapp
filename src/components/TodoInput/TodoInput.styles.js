import styled from 'styled-components'

export const Wrapper = styled.div`
  margin-top: 12px;
  width: 300px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`

export const Input = styled.input`
  font-size: 16px;
  flex-grow: 1;
  border-radius: 8px;
  border: 2px solid ${({ theme }) => theme.colors.text};
  background-color: ${({ theme }) => theme.colors.bg};
  padding: 8px;
  color: ${({ theme }) => theme.colors.text};
  outline: none;
`

export const Button = styled.button`
  margin-left: 16px;
  padding: 8px 16px;
  border: 2px solid ${({ theme }) => theme.colors.text};
  background-color: ${({ theme }) => theme.colors.bg};
  border-radius: 8px;
  color: ${({ theme }) => theme.colors.text};
  outline: none;
  cursor: pointer;
`
