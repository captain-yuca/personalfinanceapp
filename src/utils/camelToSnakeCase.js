export const camelToSnakeCase = str => str.replace(/([A-Z])/g, group => `_${group.toLowerCase()}`)
