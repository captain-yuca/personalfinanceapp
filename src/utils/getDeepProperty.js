export const getDeepProperty = (path, obj) =>
  path.split('.').reduce((prev, curr) => (prev ? prev[curr] : null), obj || null)
