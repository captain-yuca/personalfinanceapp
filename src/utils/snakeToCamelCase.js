export const snakeToCamelCase = str =>
  str.replace(/([-_][a-z])/g, group => group.toUpperCase().replace('-', '').replace('_', ''))
