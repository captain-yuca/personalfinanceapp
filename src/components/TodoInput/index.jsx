import React from 'react'
import TodoInput from './TodoInput'

function Wrapper(props) {
  return <TodoInput {...props} />
}

export default Wrapper
