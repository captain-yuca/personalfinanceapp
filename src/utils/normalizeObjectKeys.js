import { camelToSnakeCase } from './camelToSnakeCase'
import { snakeToCamelCase } from './snakeToCamelCase'

export const normalizeObjectKeys = (obj, toSnake = false) =>
  Object.entries(obj).reduce((accum, [rawKey, value]) => {
    const key = toSnake ? camelToSnakeCase(rawKey) : snakeToCamelCase(rawKey)
    return { ...accum, [key]: value }
  }, {})
