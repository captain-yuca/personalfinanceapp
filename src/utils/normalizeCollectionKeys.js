import { normalizeObjectKeys } from './normalizeObjectKeys'

export const normalizeCollectionKeys = (collection, toSnake = false) =>
  collection.map(item => normalizeObjectKeys(item, toSnake))
