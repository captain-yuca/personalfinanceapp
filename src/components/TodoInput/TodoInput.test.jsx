import React from 'react'
import { fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import uniqid from 'uniqid'
import { render } from '../../test.utils'

import TodoInput from './TodoInput'
import TodoWrapper from '.'

jest.mock('uniqid', () => () => 1)

describe('TodoInput component', () => {
  const props = {
    addTodo: jest.fn(),
  }

  it('should render correctly', () => {
    const { container, getByText } = render(<TodoInput {...props} />)

    expect(getByText('Add')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should correctly handle input change', () => {
    const { container, getByPlaceholderText } = render(<TodoInput {...props} />)

    const input = getByPlaceholderText('Enter new todo')
    userEvent.type(input, 'new element')

    expect(container).toMatchSnapshot()
  })

  it('should handle click on add btn', () => {
    const { getByPlaceholderText, getByText } = render(<TodoInput {...props} />)

    const btn = getByText('Add')
    fireEvent.click(btn)

    expect(props.addTodo).not.toBeCalled()

    const input = getByPlaceholderText('Enter new todo')
    userEvent.type(input, 'new element')

    fireEvent.click(btn)

    expect(props.addTodo).toBeCalledWith({ id: uniqid(), text: 'new element' })
  })
})

describe('TodoWrapper component', () => {
  const props = {
    addTodo: jest.fn(),
  }

  it('should render correctly', () => {
    const { container } = render(<TodoWrapper {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('should handle click on add btn', () => {
    const { getByPlaceholderText, getByText } = render(<TodoWrapper {...props} />)
    const btn = getByText('Add')
    fireEvent.click(btn)
    const input = getByPlaceholderText('Enter new todo')
    userEvent.type(input, 'new element')
    fireEvent.click(btn)
  })
})
