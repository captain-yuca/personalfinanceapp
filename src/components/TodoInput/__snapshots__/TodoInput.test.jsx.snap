// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInput component should correctly handle input change 1`] = `
.c0 {
  margin-top: 12px;
  width: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-size: 16px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  border-radius: 8px;
  border: 2px solid #ffffff;
  background-color: #212121;
  padding: 8px;
  color: #ffffff;
  outline: none;
}

.c2 {
  margin-left: 16px;
  padding: 8px 16px;
  border: 2px solid #ffffff;
  background-color: #212121;
  border-radius: 8px;
  color: #ffffff;
  outline: none;
  cursor: pointer;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Enter new todo"
      type="text"
      value="new element"
    />
    <button
      class="c2"
    >
      Add
    </button>
  </div>
</div>
`;

exports[`TodoInput component should render correctly 1`] = `
.c0 {
  margin-top: 12px;
  width: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-size: 16px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  border-radius: 8px;
  border: 2px solid #ffffff;
  background-color: #212121;
  padding: 8px;
  color: #ffffff;
  outline: none;
}

.c2 {
  margin-left: 16px;
  padding: 8px 16px;
  border: 2px solid #ffffff;
  background-color: #212121;
  border-radius: 8px;
  color: #ffffff;
  outline: none;
  cursor: pointer;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Enter new todo"
      type="text"
      value=""
    />
    <button
      class="c2"
    >
      Add
    </button>
  </div>
</div>
`;

exports[`TodoWrapper component should render correctly 1`] = `
.c0 {
  margin-top: 12px;
  width: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  font-size: 16px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  border-radius: 8px;
  border: 2px solid #ffffff;
  background-color: #212121;
  padding: 8px;
  color: #ffffff;
  outline: none;
}

.c2 {
  margin-left: 16px;
  padding: 8px 16px;
  border: 2px solid #ffffff;
  background-color: #212121;
  border-radius: 8px;
  color: #ffffff;
  outline: none;
  cursor: pointer;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Enter new todo"
      type="text"
      value=""
    />
    <button
      class="c2"
    >
      Add
    </button>
  </div>
</div>
`;
